File: LeetCode Problems/101 to 200/139. Word Break/139. Word Break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        n = len(s)
        dp = [False] * (n + 1)
        dp[n] = True
        for i in range(n, -1, -1):
            for w in wordDict:
                if (i + len(w)) <= n and s[i:i+len(w)] == w:
                    dp[i] = dp[i + len(w)]
                if(dp[i]):
                    break
        return dp[0]

File: LeetCode Problems/101 to 200/139. Word Break/139. Word Break.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string> &wordDict) {
        int n = s.length();
        vector<bool> dp(n + 1, false);
        dp[n] = true;

        for (int i = n - 1; i >= 0; i--) {
            for (const std::string &w: wordDict) {
                if (i + w.length() <= n && s.substr(i, w.length()) == w) {
                    dp[i] = dp[i + w.length()];
                }
                if (dp[i]) {
                    break;
                }
            }
        }

        return dp[0];
    }
};
